src/edit.rs

```diff
@@ -1,28 +1,69 @@
 use encoding::types::EncodingRef;
 
+use std::u32;
 use error::Error;
 use history::{Cursor, History};
 use buffer::Buffer;
 use parser::{parse, ParseError, ParseSuccess};
 use instr;
 
+#[derive(Copy,Clone)]
+pub enum EditMode {
+    Emacs,
+    Vi,
+}
+
+#[derive(Copy, Clone, PartialEq)]
+pub enum ViMode {
+    Insert,
+    Normal,
+    Replace,
+    MoveChar(instr::CharMoveType),
+    DeleteMoveChar(instr::CharMoveType),
+    ChangeMoveChar(instr::CharMoveType),
+    Delete,
+    Change,
+}
+
+/// Set a new vi mode based on the current vi mode.
+///
+/// There are several common mode transitions that we handle here.
+macro_rules! set_next_vi_mode {
+    ( $ctx:ident ) => {
+        $ctx.vi_mode = match $ctx.vi_mode {
+            ViMode::Delete => ViMode::Normal,
+            ViMode::Change => ViMode::Insert,
+            ViMode::DeleteMoveChar(_) => ViMode::Normal,
+            ViMode::ChangeMoveChar(_) => ViMode::Insert,
+            _ => ViMode::Normal,
+        }
+    }
+}
+
 pub struct EditCtx<'a> {
     buf: Buffer,
     history_cursor: Cursor<'a>,
     prompt: &'a str,
     seq: Vec<u8>,
-    enc: EncodingRef
+    enc: EncodingRef,
+    mode: EditMode,
+    vi_mode: ViMode,
+    vi_count: u32,
```

Comment:
I thought about something like this, but I didn't realize I could have an impl for an enum type. I'll make the change.